### PR TITLE
fix(crystallization): YAML multiline patch + H1 rename (closes #1427)

### LINKED ISSUE #1427: Epic: Draft YAML patches and H1 overrides (supersedes #1369, #1368)
## Combined coder briefing

_Structured extracts from each superseded issue (plus bold **Description** / **Impact** blocks when issues do not use `## Summary`). Verbatim originals remain under **Reference**._

### #1369: applyOverridesToDraft regex strictness prevents H1 updates
**Summary / description**

In `extensions/memory-hybrid/services/crystallization-proposer.ts`, the H1 replacement regex assumes the H1 exactly matches `proposal.skillName`.

**Impact**

If a user manually edits the draft's H1 heading to something else (e.g., `# My Custom Skill`) before approving with a name override, the regex won't match, and the H1 won't be updated to the new overridden name.

---

### #1368: patchOpeningYamlField multi-line value corruption
**Summary / description**

In `extensions/memory-hybrid/services/crystallization-proposer.ts`, the regex used to patch YAML fields only matches the first line of a value:
```typescript
const re = new RegExp(`^${escapeRegExp(key)}:\\s*.*$`, "m");
```

**Impact**

If the YAML frontmatter contains a multi-line value (e.g., `description: >\n  Line 1\n  Line 2`), replacing it will leave the subsequent lines dangling, corrupting the YAML structure.

---

## Quick links (superseded issues)

- https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1369
- https://github.com/markus-lassfolk/openclaw-hybrid-memory/issues/1368

## Supersedes
- #1369, #1368

---

## Reference: merged detail from superseded issues

_Verbatim closed-issue bodies (expand any row):_

<details>
<summary><strong>#1369: applyOverridesToDraft regex strictness prevents H1 updates</strong></summary>

**Description**
In `extensions/memory-hybrid/services/crystallization-proposer.ts`, the H1 replacement regex assumes the H1 exactly matches `proposal.skillName`.

**Impact**
If a user manually edits the draft's H1 heading to something else (e.g., `# My Custom Skill`) before approving with a name override, the regex won't match, and the H1 won't be updated to the new overridden name.

</details>

<details>
<summary><strong>#1368: patchOpeningYamlField multi-line value corruption</strong></summary>

**Description**
In `extensions/memory-hybrid/services/crystallization-proposer.ts`, the regex used to patch YAML fields only matches the first line of a value:
```typescript
const re = new RegExp(`^${escapeRegExp(key)}:\\s*.*$`, "m");
```

**Impact**
If the YAML frontmatter contains a multi-line value (e.g., `description: >\n  Line 1\n  Line 2`), replacing it will leave the subsequent lines dangling, corrupting the YAML structure.

</details>

### PR DESCRIPTION
**Pull request by markus-lassfolk**
## Summary
- **`patchOpeningYamlField`**: Replaces the full YAML value for a key, including literal/folded block scalars (`|`, `>` and common chomping/indent indicators) up to the next top-level key, plus indented plain multiline continuations. Bounded scan (512 lines). Preserves CRLF inside frontmatter when present. Values with special characters use JSON-style quoting.
- **Rename / approve overrides**: Replaces the first Markdown ATX H1 in the body after frontmatter when present, instead of only rewriting a heading that exactly matched the old `proposal.skillName`.

## Tests
- Unit tests for `patchOpeningYamlField` (literal block, folded block, indented continuation, CRLF).
- Integration-style test for `approveProposal` with a custom H1 title and rename.

Closes #1427.

Made with [Cursor](https://cursor.com)

---

> [!NOTE]
> **Medium Risk**
> Touches crystallization install/validation flow and persistence by adding a new lifecycle state and rewriting frontmatter/H1 mutation logic, which could affect how skills are written and validated. Main risk is unexpected content rewriting or status transitions for installed skills during rescans.
> 
> **Overview**
> Improves `approveProposal` rename behavior by updating the *first* body H1 after frontmatter (CRLF-safe) and strengthening `patchOpeningYamlField` to correctly replace **multiline YAML values** (literal/folded block scalars and indented continuations) while preserving existing line endings.
> 
> Extends proposal lifecycle with a new `quarantined` status, adds `CrystallizationStore.quarantine()`, and introduces `CrystallizationProposer.rescanInstalledSkills()` to re-validate on-disk installed `SKILL.md` files, persist fresh validation results, and quarantine installs that now fail validation.
> 
> Adds comprehensive tests covering multiline YAML patching edge cases, rename/H1 rewriting, and legacy detection when skills are prefixed with HTML comments.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8607cf675701d792060169fd11a283c583dbf77d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: extensions/memory-hybrid/services/crystallization-proposer.ts [hunk 1]
```diff
@@ -336,6 +366,71 @@ export class CrystallizationProposer {
...
+        if (validation.approvalDecision === "deny") {
+          const detail = detailSkillProposalValidation(validation);
+          const reason = `stale validation: ${detail}`;
+          const updated = this.crystallizationStore.quarantine(proposal.id, reason);
+          if (updated) {
+            quarantined++;
+            messages.push(`Quarantined ${proposal.skillName}: ${summarizeSkillProposalValidation(validation)}`);
+          } else {
+            errors.push(`${proposal.id}: quarantine update failed`);
+          }
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/crystallization-proposer.ts:497]: Fixed in 754a8e59aa531d2cdd3dcd7534a54ffa847adb9c — added the missing crystallization store quarantined lifecycle status and quarantine() transition for rescanInstalledSkills.

File: extensions/memory-hybrid/services/crystallization-proposer.ts [hunk 2]
```diff
@@ -442,22 +534,113 @@ ${proposal.skillContent}`;
...
+    const ch = fragment[i]!;
+    if (ch === "'" && !inDouble) inSingle = !inSingle;
+    else if (ch === '"' && !inSingle) inDouble = !inDouble;
+    else if (ch === "#" && !inSingle && !inDouble) return fragment.slice(0, i).trimEnd();
+  }
+  return fragment.trimEnd();
+}
+
+function isTopLevelYamlKeyLine(line: string): boolean {
+  return /^[A-Za-z_][A-Za-z0-9_-]*:\s/.test(line);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/crystallization-proposer.ts]: Fixed in 754a8e59aa531d2cdd3dcd7534a54ffa847adb9c — top-level YAML key detection now accepts a colon at end-of-line, preserving keys like tags: after block scalar replacement.

File: extensions/memory-hybrid/tests/crystallization-proposer.test.ts [hunk 3]
```diff
@@ -355,3 +356,129 @@ describe("CrystallizationProposer.rejectProposal", () => {
...
+function skillFixture(skillName: string, h1Title: string): string {
+  return `---
+name: ${skillName}
+description: Short description for validation.
+category: testing
+---
+
+# ${h1Title}
+
+${MINIMAL_VALID_SKILL_SECTIONS}
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/tests/crystallization-proposer.test.ts:553]: Fixed in 754a8e59aa531d2cdd3dcd7534a54ffa847adb9c — updated the approval fixture with required provenance, verification, anti-patterns, and concrete examples so the H1 rename test exercises the validator path successfully.

File: extensions/memory-hybrid/services/crystallization-proposer.ts [hunk 4]
```diff
@@ -442,22 +534,113 @@ ${proposal.skillContent}`;
...
+}
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/crystallization-proposer.ts:696]: Fixed in 754a8e59aa531d2cdd3dcd7534a54ffa847adb9c — top-level YAML key detection now accepts a colon at end-of-line, preserving keys like tags: after block scalar replacement.

File: extensions/memory-hybrid/services/crystallization-proposer.ts [hunk 5]
```diff
@@ -442,22 +536,145 @@ ${proposal.skillContent}`;
...
+  if (blockHdr) {
+    const afterMarker = (blockHdr[2] ?? "").trim();
+    if (/^([-+]|[1-9][0-9]*)?$/.test(afterMarker)) {
+      let j = startIdx + 1;
+      while (j < lines.length && j - startIdx < MAX_YAML_VALUE_SCAN_LINES) {
+        if (isTopLevelYamlKeyLine(lines[j]!)) break;
+        j++;
+      }
+      return j;
+    }
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/crystallization-proposer.ts:719]: Fixed in 754a8e59aa531d2cdd3dcd7534a54ffa847adb9c — YAML block scalar indicator handling now accepts combined indent/chomping forms such as |2- before scanning the block.

File: extensions/memory-hybrid/services/crystallization-proposer.ts [hunk 6]
```diff
@@ -24,6 +24,30 @@ import {
...
+    body = body.slice(fmMatch[0].length);
+  }
+  const leadingNewlineMatch = body.match(/^\r?\n/);
+  const leadingNewline = leadingNewlineMatch ? leadingNewlineMatch[0] : "";
+  body = body.replace(/^\r?\n/, "");
+  const h1Line = /^#[ \t]+(?!#)\S.*$/m;
+  if (!h1Line.test(body)) {
+    return skillContent;
+  }
+  return head + leadingNewline + body.replace(h1Line, `# ${newTitle}`);
```
[the remark this answers is not in the document]
Comment (markus-lassfolk) [reply extensions/memory-hybrid/services/crystallization-proposer.ts]: Fixed in 754a8e59aa531d2cdd3dcd7534a54ffa847adb9c — H1 replacement now excludes line terminators from the match and reinserts the original CR when present; added CRLF coverage.

**Comment by markus-lassfolk** [comment]
@copilot please fix the merge conflict for this PR. The branch is behind main and needs a rebase/merge. 

Important:
- Rebase or merge main into the PR branch to resolve conflicts.
- Commit directly to the PR branch.
- Do not open a new PR.
- After fixing, leave a short summary of what was changed.

Repository: markus-lassfolk/openclaw-hybrid-memory
PR: #1457
Current head: 0a962378040e792391f93de933df05d3e6c3ea2a
Branch: feat/issue-1427-yaml-h1-draft

**Comment by markus-lassfolk** [comment]
Stewardship pulse 2026-05-16T08:36Z: PR advanced to stage/council. CI green, mergeable, fixes-needed cleared. Council/deep verification now required for merge readiness. Head: a468ebb2f6726ded752afc7ffc32e20a6a1eab39, idle=12596s, next_owner=Council/deep_verifier.

**Comment by markus-lassfolk** [comment]
update-behind executed; new head SHA pending from GitHub after branch refresh.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1457
Head SHA: a468ebb2f6726ded752afc7ffc32e20a6a1eab39 (pre-update; new head pending)
stage: stage/council (maintained)
blockers: none
action_taken: gate-managed --update-behind; branch refreshed against main; waiting for new head + CI
next_owner: wait for new head SHA + checks settle; then re-triage for Council/dispatch or stage/approved
proof: https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1457

**Comment by markus-lassfolk** [comment]
@copilot please repair this PR branch and push to the existing branch.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1457 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1457
Title: fix(crystallization): YAML multiline patch + H1 rename (closes #1427)
Head SHA: 94e818efa46493089e653b22fdbac3c9020c8879
Branch: feat/issue-1427-yaml-h1-draft
Current merge state: CONFLICTING/DIRTY
Labels: extensions,size/XL,stage/council

Requested work:
- Resolve merge conflicts / dirty branch state or failing CI blockers on this PR head.
- Push to the existing PR branch; do not open a new PR.
- Preserve this PR's feature intent and current main behavior.
- Report what was fixed, commit SHA, and any remaining blockers.

next_owner=Copilot; wait_until=Copilot push/comment or timeout.

**Comment by markus-lassfolk** [comment]
@copilot please repair this PR branch and push to the existing branch.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1457 https://github.com/markus-lassfolk/openclaw-hybrid-memory/pull/1457
Title: fix(crystallization): YAML multiline patch + H1 rename (closes #1427)
Head SHA: e00407b50348f5034d0674cadf2ca4b5e301af6d
Branch: feat/issue-1427-yaml-h1-draft
Current merge state: CONFLICTING/DIRTY
Labels: extensions,size/XL,stage/council

Requested work:
- Resolve merge conflicts / dirty branch state or failing CI blockers on this PR head.
- Push to the existing PR branch; do not open a new PR.
- Preserve this PR's feature intent and current main behavior.
- Report what was fixed, commit SHA, and any remaining blockers.

next_owner=Copilot; wait_until=Copilot push/comment or timeout.

**Comment by markus-lassfolk** [comment]
@copilot please fix this bounded blocker on the current PR head.

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1457
Head: 8607cf675701d792060169fd11a283c583dbf77d
Blockers:
- Test (Node 22) FAILURE — https://github.com/markus-lassfolk/openclaw-hybrid-memory/actions/runs/25989145567/job/76392014638
- Test Coverage FAILURE — https://github.com/markus-lassfolk/openclaw-hybrid-memory/actions/runs/25989145567/job/76392014665
Expected fix: Fix failing tests and ensure test coverage passes.
Verification: Test (Node 22) and Test Coverage pass on re-run.

Stewardship next_owner=Copilot; attempt=1/2; will recheck after verification trigger.

**Comment by markus-lassfolk** [comment]
## Council Review Request

Repo/PR: markus-lassfolk/openclaw-hybrid-memory#1457
Head SHA: 3f045282be8b5b4233f58f2118025002e04a944a
Stage: stage/review (promotion candidate)
Title: fix(crystallization): YAML multiline patch + H1 rename (closes #1427)

Acceptance criteria: Fix for #1427 — YAML multiline patch + H1 rename. CI/checks all green, no unresolved threads, no blockers.

Current state: CI=green (25 checks passed), threads=0 unresolved, labels=stage/review + extensions + size/XL (fixes-needed removed — previous blockers resolved).

Review request: Verify current head against issue #1427 acceptance criteria and repo policy. Post APPROVED or CHANGES_REQUESTED with exact blockers tied to this head SHA.
